src/field/scrcmd_mystery_gift.c

```diff
@@ -0,0 +1,684 @@
+    SetMonData(pokemon, MON_DATA_GENDER, &gender);
+
+    gender = GetMonData(pokemon, MON_DATA_HP_IV, NULL)
+        + GetMonData(pokemon, MON_DATA_ATK_IV, NULL)
+        + GetMonData(pokemon, MON_DATA_DEF_IV, NULL)
+        + GetMonData(pokemon, MON_DATA_SPEED_IV, NULL)
+        + GetMonData(pokemon, MON_DATA_SPATK_IV, NULL)
+        + GetMonData(pokemon, MON_DATA_SPDEF_IV, NULL);
+    if (gender == 0) {
+        u16 ivRand = LCRandom();
+        rand = ivRand & 0x1Fu;
+        SetMonData(pokemon, MON_DATA_HP_IV, &rand);
+        rand = (ivRand & 0x3E0u) >> 5;
+        SetMonData(pokemon, MON_DATA_ATK_IV, &rand);
+        rand = (ivRand & 0x7C00u) >> 10;
+        SetMonData(pokemon, MON_DATA_DEF_IV, &rand);
+        ivRand = LCRandom();
+        rand = ivRand & 0x1Fu;
+        SetMonData(pokemon, MON_DATA_SPEED_IV, &rand);
+        rand = (ivRand & 0x3E0u) >> 5;
+        SetMonData(pokemon, MON_DATA_SPATK_IV, &rand);
+        rand = (ivRand & 0x7C00u) >> 10;
+        SetMonData(pokemon, MON_DATA_SPDEF_IV, &rand);
+    }
+
+    SaveSpecialRibbons* ribbons = Save_SpecialRibbons_Get(fieldSys->saveData);
+    if (GetMonData(pokemon, MON_DATA_RED_RIBBON, NULL)) {
+        ribbons->ribbons[GetSpecialRibbonNo(RIBBON_RED)] = srcRibbons[0];
+    }
+    if (GetMonData(pokemon, MON_DATA_GREEN_RIBBON, NULL)) {
+        ribbons->ribbons[GetSpecialRibbonNo(RIBBON_GREEN)] = srcRibbons[1];
+    }
+    if (GetMonData(pokemon, MON_DATA_BLUE_RIBBON, NULL)) {
+        ribbons->ribbons[GetSpecialRibbonNo(RIBBON_BLUE)] = srcRibbons[2];
+    }
+    if (GetMonData(pokemon, MON_DATA_FESTIVAL_RIBBON, NULL)) {
+        ribbons->ribbons[GetSpecialRibbonNo(RIBBON_FESTIVAL)] = srcRibbons[3];
+    }
+    if (GetMonData(pokemon, MON_DATA_CARNIVAL_RIBBON, NULL)) {
+        ribbons->ribbons[GetSpecialRibbonNo(RIBBON_CARNIVAL)] = srcRibbons[4];
+    }
+    if (GetMonData(pokemon, MON_DATA_CLASSIC_RIBBON, NULL)) {
+        ribbons->ribbons[GetSpecialRibbonNo(RIBBON_CLASSIC)] = srcRibbons[5];
+    }
+    if (GetMonData(pokemon, MON_DATA_PREMIER_RIBBON, NULL)) {
+        ribbons->ribbons[GetSpecialRibbonNo(RIBBON_PREMIER)] = srcRibbons[6];
+    }
+    if (GetMonData(pokemon, MON_DATA_HOENN_MARINE_RIBBON, NULL)) {
+        ribbons->ribbons[GetSpecialRibbonNo(RIBBON_HOENN_MARINE)] = srcRibbons[7];
+    }
+    if (GetMonData(pokemon, MON_DATA_HOENN_LAND_RIBBON, NULL)) {
+        ribbons->ribbons[GetSpecialRibbonNo(RIBBON_HOENN_LAND)] = srcRibbons[8];
+    }
+    if (GetMonData(pokemon, MON_DATA_HOENN_SKY_RIBBON, NULL)) {
+        ribbons->ribbons[GetSpecialRibbonNo(RIBBON_HOENN_SKY)] = srcRibbons[9];
+    }
+
+    if (mgData->fixedOT == OT_ID_PLAYER_ID) {
+        String* playerName = PlayerProfile_GetPlayerName_NewString(profile, HEAP_ID_32);
+        u32 trainerId = PlayerProfile_GetTrainerID(profile);
+        BOOL gender = PlayerProfile_GetTrainerGender(profile);
+
+        tmpPokemon = AllocMonZeroed(HEAP_ID_32);
+#ifdef UBFIX
+        GF_ASSERT(pokemon != NULL);
```

Comment:
why would this fix UB? it'd just error if it was null, and not actually fix the issue

Reply:
if there's a memory allocation error, nothing you do will let the game continue to run in a valid state. the best course of action is to error out.